src/reference_implementations/prompting_vector_llms/prompt_ensembling/bootstrap_ensembling.ipynb

```diff
@@ -0,0 +1,588 @@
+    "\n",
+    "all_llama_responses = []\n",
+    "\n",
+    "number_of_voters = 7\n",
+    "for voter_number in range(number_of_voters):\n",
+    "    print(f\"Starting MC Choice Response Number: {voter_number + 1}\")\n",
+    "    demonstration_pool = copa_data_set[0:demonstration_candidates]\n",
+    "    demonstrations = sample(demonstration_pool, n_demonstrations)\n",
+    "    prompts = []\n",
+    "    labels = []\n",
+    "    int_labels = []\n",
+    "    choices = []\n",
+    "    for premise, label, phrase, first_choice, second_choice in test_pool:\n",
+    "        int_labels.append(label)\n",
+    "        choices.append((first_choice, second_choice))\n",
+    "        labels.append(create_mc_prompt_answer(label))\n",
+    "        prompts.append(create_mc_prompt(demonstrations, premise, phrase, first_choice, second_choice))\n",
+    "\n",
+    "    llama_responses = []\n",
+    "    prompt_batches = split_prompts_into_batches(prompts, 10)\n",
+    "    for prompt_batch in tqdm(prompt_batches):\n",
+    "        generations = llama_model.generate(prompt_batch, llama_generation_config)\n",
+    "        llama_responses.extend(process_mc_generation_text(generations.generation[\"sequences\"]))\n",
+    "\n",
+    "    all_llama_responses.append(llama_responses)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Accuracy: 0.75\n"
+     ]
+    }
+   ],
+   "source": [
+    "total = 0\n",
+    "correct = 0\n",
+    "\n",
+    "for response_tuple, label in zip(zip(*all_llama_responses), labels):\n",
+    "    predicted_label = Counter(response_tuple).most_common(1)[0][0]\n",
+    "    total += 1\n",
+    "    if predicted_label == label:\n",
+    "        correct += 1\n",
+    "\n",
+    "print(f\"Accuracy: {correct/total}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We actually get a pretty nice increase in the accuracy here (it was only 0.68 in the other notebook). The quality and distribution of the demonstrations has an impact on how well a model performs a task. By sampling several sets of demonstrations, we were able to improve our results!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Generation Formulation\n",
+    "\n",
+    "We'll use Falcon for this example as an alternative to using LLaMA-2 above. It performed a little better in the previous notebook example as well. In this setting, we're going to use the same set of demonstrations, but we'll allow the model to sample different generation trajectories and see what happens when we use a simple voting strategy to tabulate the prediction. Generation is a bit more \"expensive\" than the multiple choice setting. So we'll perform 3 distinct generations and see if we can get any improvement."
+   ]
```

Comment:
Would using different sets demonstrations like we did in the MC formulation here? I think it makes sense to keep this as it is since it showcases the effects of only relying on the stochasticity of generation trajectories, but just wondering if you've tried it

Reply:
I think that would definitely make sense. I didn't try it and/or report it here, because I was trying to showcase a different possible idea. However, I'll put in a comment stating that one might try demonstration re-sampling for this problem as well.